src/types/adapters/IntegrationAdapter.ts

```diff
@@ -2,43 +2,39 @@ import { assertNever } from 'assert-never';
 
 import { Schemas } from '../../generated/OpenapiIntegrations';
 import {
+    CamelIntegrationType,
     Integration,
     IntegrationBase,
-    IntegrationCamel, IntegrationEmailSubscription,
+    IntegrationCamel,
+    IntegrationEmailSubscription,
     IntegrationHttp,
-    IntegrationType,
+    IntegrationType, isCamelType,
     NewIntegration,
     ServerIntegrationRequest,
     ServerIntegrationResponse
 } from '../Integration';
 
-const getIntegrationType = (type: Schemas.EndpointType | undefined): IntegrationType => {
-    switch (type) {
-        case Schemas.EndpointType.Enum.camel:
-            return IntegrationType.CAMEL;
-        case Schemas.EndpointType.Enum.webhook:
-            return IntegrationType.WEBHOOK;
-        case Schemas.EndpointType.Enum.email_subscription:
-            return IntegrationType.EMAIL_SUBSCRIPTION;
-        case Schemas.EndpointType.Enum.default:
-        case undefined:
-            throw new Error(`Unexpected type: ${type}`);
-        default:
-            assertNever(type);
+interface ExternalCompositeTyped {
+    type: string;
+    sub_type?: string | null;
+}
+
+export const getIntegrationType = (serverIntegration: ExternalCompositeTyped): IntegrationType => {
+    for (const integration of Object.values(IntegrationType)) {
+        if (serverIntegration.sub_type) {
+            if (integration === `${serverIntegration.type}:${serverIntegration.sub_type}`) {
+                return integration as IntegrationType;
+            }
+        } else if (integration === serverIntegration.type) {
+            return integration as IntegrationType;
+        }
     }
+
+    throw new Error(`Unexpected type: ${serverIntegration.type} with subtype: ${serverIntegration.sub_type}`);
 };
 
-export const getEndpointType = (type: IntegrationType): Schemas.EndpointType => {
-    switch (type) {
-        case IntegrationType.WEBHOOK:
-            return Schemas.EndpointType.Enum.webhook;
-        case IntegrationType.CAMEL:
-            return Schemas.EndpointType.Enum.camel;
-        case IntegrationType.EMAIL_SUBSCRIPTION:
-            return Schemas.EndpointType.Enum.email_subscription;
-        default:
-            assertNever(type);
-    }
+export const getEndpointType = (type: IntegrationType): [ Schemas.EndpointType, string] => {
+    return type.split(':', 2) as [ Schemas.EndpointType, string ];
```

Comment:
Suggestion: Instead of a custom data structure holding two values, can it be instead split into two separate variables? Either by passing them separately or passing an object (`endpoint_type: {type: String, sub_type: String}`?

Reply:
Sure. Actually this is an internal method (will remove the `export`)

Reply:
Where does the `type` value come from? Is that also internal?

Reply:
Yes, it's internal to the UI code: see: https://github.com/RedHatInsights/notifications-frontend/blob/master/src/types/Integration.ts#L6-L10
It converts that type to a type/subType that is consumed by the notifications-backend

Reply:
Oh, so the origin is the constant? If so, then the constant might have it as a struct.
Type-wise something like this, changing the enum into union of several types.
```
export const SPLUNK = Object.freeze({ type: 'camel', sub_type: 'splunk'  });
type IntegrationType = | WEBHOOK | EMAIL_SUBSCRIPTION | SPLUNK;
```

Reply:
That looks nice! I'll play with that idea! Thanks!

Reply:
Played with this a bit, but ran into some issues where a enum is far easier to manage. (narrowing was one of these issues, as the enum is used with some cases/assertNevers) - I might revisit it again to have a 1-1 match between the server/client.

Reply:
No worries. My first thought was whether the enum can hold an object, which it probably cannot. For a listing possible types there would have to be additional array (if TS does not allow unwrapping the type definition).